### PR TITLE
feat(module: datepicker): add 12-hour time support

### PR DESCRIPTION
**Pull request by Alexbits**
Add 12-hour time support to the DatePicker/TimePicker/RangePicker
### 🤔 This is a ...

- [x] New feature
- [x] Bug fix
- [x] Site / documentation update
- [x] Demo update
- [x] Component style update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by ElderJames** [comment]
Thanks for contribute this @Alexbits , this make a great help!

In terms of globalization, I think AM and PM can be used.  NET built-in culture of the corresponding string.
```
  CultureInfo.CurrentCulture.DateTimeFormat.AMDesignator;
  CultureInfo.CurrentCulture.DateTimeFormat.PMDesignator
```

**Comment by Alexbits** [comment]
> Thanks for contribute this @Alexbits , this make a great help!
> 
> In terms of globalization, I think AM and PM can use . NET built-in string for the current culture as the fallback.
> 
> ```
>   CultureInfo.CurrentCulture.DateTimeFormat.AMDesignator;
>   CultureInfo.CurrentCulture.DateTimeFormat.PMDesignator
> ```

Thank you for your work  @ElderJames ! I thought about this, but I was unsure if it was worth it since you provided the strings in the DateLocale class.

Where do you mean to implement this fallback?

**Comment by ElderJames** [comment]
Just the text in the input component, it can't be localized.

This is en-US:
image

This is zh-CN, you can switch the language on the right-top of the site.
image

File: components/locales/en-US.json [hunk 1]
```diff
@@ -41,7 +41,7 @@
       "yearFormat": "yyyy",
       "dateFormat": "yyyy-MM-dd",
       "dayFormat": "d",
-      "dateTimeFormat": "yyyy-MM-dd HH:mm:ss",
+      "dateTimeFormat": "yyyy-MM-dd hh:mm:ss tt",
```
Comment (ElderJames) [inline components/locales/en-US.json:44]: Our default is 24-hours time format.

File: components/locales/en-US.json [hunk 2]
```diff
@@ -103,7 +103,7 @@
       "yearFormat": "yyyy",
       "dateFormat": "M/d/yyyy",
       "dayFormat": "d",
-      "dateTimeFormat": "M/d/yyyy HH:mm:ss",
+      "dateTimeFormat": "M/d/yyyy hh:mm:ss tt",
```
Comment (ElderJames) [inline components/locales/en-US.json:106]: Our default is 24-hours time format.

File: components/locales/en-US.json (hunk 1 above)
Comment (Alexbits) [reply components/locales/en-US.json:44]: ok

**Comment by Alexbits** [comment]
Hi @ElderJames  I checked the original ant.design, and AM/PM in the selection panel is not localized 🤔

image

**Comment by ElderJames** [comment]
There are a issue when I press the once of these buttons and they would switch by turns.
timepicker-12hours

**Comment by Alexbits** [comment]
> There are an issue when I press the one of these buttons and they would switch by turns. timepicker-12hours [ ](https://user-images.githubusercontent.com/7550366/172441905-ef264356-17c3-4ed3-a1e3-633fe1182ba5.gif) timepicker-12hours [ ](https://user-images.githubusercontent.com/7550366/172441905-ef264356-17c3-4ed3-a1e3-633fe1182ba5.gif)

Fixed. Sorry! I've missed this somehow.

**Review by ElderJames — DISMISSED** [review]
LGTM. Thanks.

**Comment by ElderJames** [comment]
For DatePicker, the display format in input has no AM/PM.

image

**Comment by Alexbits** [comment]
> For DatePicker, the display format in input has no AM/PM.
Fixed

File: components/date-picker/internal/DatePickerBase.cs [hunk 3]
```diff
@@ -658,7 +666,7 @@ private string InternalFormat
                             DatePickerType.Date => GetTimeFormat(),
                             DatePickerType.Month => Locale.Lang.YearMonthFormat,
                             DatePickerType.Year => Locale.Lang.YearFormat,
-                            DatePickerType.Time => Locale.Lang.TimeFormat,
+                            DatePickerType.Time => Use12Hours ? Locale.Lang.TimeFormat12Hour : Locale.Lang.TimeFormat,
```
Comment (ElderJames) [inline components/date-picker/internal/DatePickerBase.cs:669]: We can use `private TimeFormat =>  Use12Hours ? Locale.Lang.TimeFormat12Hour : Locale.Lang.TimeFormat` to reduce duplication.

File: components/locales/en-US.json (hunk 2 above)
Comment (ElderJames) [inline components/locales/en-US.json:106]: `dateTimeFormat` have to be 24-hours, need to add `timeFormat12Hour` and  `dateTimeFormat12Hour`.

File: components/date-picker/internal/DatePickerDatetimePanel.razor [hunk 4]
```diff
@@ -71,23 +71,25 @@
...
+        if (Value is null)
+        {
+            localValue = "";
+            selected = (_, _) => "";
+        }
+        else
+        {
+            localValue = Value.Value.ToString(timeFormat);
+            selected = (viewTime, valueTime) => viewTime == valueTime ? $"{PrefixCls}-time-panel-cell-selected" : "";
+        }
```
Comment (ElderJames) [inline components/date-picker/internal/DatePickerDatetimePanel.razor:89]: Please fix the indent.

Comment (Alexbits) [reply components/date-picker/internal/DatePickerDatetimePanel.razor:89]: Ok. Do you mean it should be indented like below? 

image

Comment (ElderJames) [reply components/date-picker/internal/DatePickerDatetimePanel.razor:89]: Oh no, the razor files should be indent with two spaces, just follow the default setting of blazor.

File: components/date-picker/internal/DatePickerBase.cs (hunk 3 above)
Comment (Alexbits) [reply components/date-picker/internal/DatePickerBase.cs:669]: Please check below. It looks better, and no need for private property. What do you think?

image

Comment (ElderJames) [reply components/date-picker/internal/DatePickerBase.cs:669]: That's fine 👍

Comment (ElderJames) [reply components/date-picker/internal/DatePickerBase.cs:669]: I see the same line here
image

File: components/locales/en-US.json (hunk 2 above)
Comment (Alexbits) [reply components/locales/en-US.json:106]: This should be added to DatePicker/Calendar in every JSON locale file, not only en-US.json? Right?

File: components/date-picker/internal/DatePickerDatetimePanel.razor (hunk 4 above)
Comment (Alexbits) [reply components/date-picker/internal/DatePickerDatetimePanel.razor:89]: It was like this before I changed it. That's why I'm asking what indentation you mean.

File: components/date-picker/internal/DatePickerBase.cs (hunk 3 above)
Comment (Alexbits) [reply components/date-picker/internal/DatePickerBase.cs]: Committed just now

File: components/date-picker/internal/DatePickerDatetimePanel.razor (hunk 4 above)
Comment (Alexbits) [reply components/date-picker/internal/DatePickerDatetimePanel.razor:89]: I'm not sure why but Visual Studio formats it like it was in my commit. I use ctrl+k+d and default settings.

Comment (Alexbits) [reply components/date-picker/internal/DatePickerDatetimePanel.razor:89]: It looks like the standard indentation in VS 2022 is one TAB. It formats the lines you highlighted with two TABs to align with the previous code. The `<div class='@($"{PrefixCls}-datetime-panel")'` makes this indentation. Please check it in Visual Studio. It looks like the default formatting to me.

**Comment by Alexbits** [comment]
I also see another issue with the TimePicker. The selected time is often not visible. The TimePicker should auto-scroll itself. 🤔

File: components/locales/en-US.json (hunk 2 above)
Comment (ElderJames) [reply components/locales/en-US.json]: Yes, but it can fallback to en-US using the default values of the Locale Class.

Comment (Alexbits) [reply components/locales/en-US.json]: Ok. I've added it to the en-US.

File: components/date-picker/internal/DatePickerDatetimePanel.razor (hunk 4 above)
Comment (ElderJames) [reply components/date-picker/internal/DatePickerDatetimePanel.razor:89]: Everything ok in vs diff 😀

**Comment by ElderJames** [comment]
> I also see another issue with the TimePicker. The selected time is often not visible. The TimePicker should auto-scroll itself when the selected value is not visible. 🤔
Yes, antd is designed to scroll to the top when time item clicked.  We're not there yet. Do you have any ideas?

**Comment by Alexbits** [comment]
> > I also see another issue with the TimePicker. The selected time is often not visible. The TimePicker should auto-scroll itself when the selected value is not visible. 🤔
> 
> Yes, antd is designed to scroll to the top when time item clicked. We're not there yet. Do you have any ideas?

I'm looking into it.

**Review by ElderJames — APPROVED** [review]

**Pull request merged**